File: Documentation/5.4/Samples/python/ClientApi/Session/Querying/HowToPerformGroupByQuery.py
from __future__ import annotations
from typing import Dict, Any, List

from ravendb import GroupByField
from ravendb.infrastructure.orders import Order, Address, OrderLine

from examples_base import ExampleBase

from ravendb.documents.queries.group_by import GroupBy


class CountryAndQuantity:
    def __init__(self, country: str = None, ordered_quantity: int = None):
        self.country = country
        self.ordered_quantity = ordered_quantity


class EmployeeAndCompany:
    def __init__(self, employee: str = None, company: str = None):
        self.employee = employee
        self.company = company

    @classmethod
    def from_json(cls, json_dict: Dict[str, Any]) -> EmployeeAndCompany:
        return cls(json_dict["employee"], json_dict["company"])


class CountByCompanyAndEmployee:
    def __init__(self, count: int = None, company: str = None, employee_identifier: str = None):
        self.count = count
        self.company = company
        self.employee_identifier = employee_identifier


class CountOfEmployeeAndCompanyPairs:
    def __init__(self, employee_company_pair: EmployeeAndCompany, count: int):
        self.employee_company_pair = employee_company_pair
        self.count = count

    @classmethod
    def from_json(cls, json_dict: Dict[str, Any]) -> CountOfEmployeeAndCompanyPairs:
        return cls(EmployeeAndCompany.from_json(json_dict["employee_company_pair"]), json_dict["count"])


class ProductInfo:
    def __init__(self, count: int, product: str, quantity: int):
        self.count = count
        self.product = product
        self.quantity = quantity

    @classmethod
    def from_json(cls, json_dict: Dict[str, Any]) -> ProductInfo:
        return cls(json_dict["count"], json_dict["product"], json_dict["quantity"] if "quantity" in json_dict else 0)


class ProductsInfo:
    def __init__(self, count: int, products: List[str], quantity: int = None):
        self.count = count
        self.products = products
        self.quantity = quantity

    @classmethod
    def from_json(cls, json_dict: Dict[str, Any]) -> ProductsInfo:
        return cls(json_dict["count"], json_dict["products"], json_dict["quantity"] if "quantity" in json_dict else 0)


class HowToPerformGroupByQuery(ExampleBase):
    def setUp(self):
        super().setUp()
        with self.embedded_server.get_document_store("GroupByQuery") as store:
            with store.open_session() as session:
                session.store(
                    Order(
                        "Funny Order",
                        "companies/1",
                        "employees/1",
                        ship_to=Address(country="Chad"),
                        lines=[
                            OrderLine("products/1", "iPhone 15", 199.9, 20),
                            OrderLine("products/2", "Apple Vision Pro", 6666, 2),
                        ],
                    )
                )
                session.store(
                    Order(
                        "Even More Funny Order",
                        "companies/1",
                        "employees/3",
                        ship_to=Address(country="Poland"),
                        lines=[
                            OrderLine("products/3", "Grain", 2, 5000),
                            OrderLine("products/1", "iPhone 15", 1999.9, 500),
                        ],
                    )
                )
                session.save_changes()

    def test_how_to_perform_group_by_query(self):
        with self.embedded_server.get_document_store("GroupByQuery") as store:
            with store.open_session() as session:
                # region group_by_1
                orders = list(
                    session.query(object_type=Order)
                    .group_by("ship_to.country")
                    .select_key("ship_to.country", "country")
                    .select_sum(GroupByField("lines[].quantity", "ordered_quantity"))
                    .of_type(CountryAndQuantity)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_2
                results = list(
                    session.query(object_type=Order)
                    .group_by("employee", "company")
                    .select_key("employee", "employee_identifier")
                    .select_key("company")
                    .select_count()
                    .of_type(CountByCompanyAndEmployee)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_3
                orders = list(
                    session.query(object_type=Order)
                    .group_by("employee", "company")
                    .select_key("key()", "employee_company_pair")
                    .select_count("count")
                    .of_type(CountOfEmployeeAndCompanyPairs)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_4
                products = list(
                    session.query(object_type=Order)
                    .group_by(GroupBy.array("lines[].product"))
                    .select_key("key()", "products")
                    .select_count()
                    .of_type(ProductsInfo)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_5
                products = list(
                    session.advanced.document_query(object_type=Order)
                    .group_by("lines[].product", "ship_to.country")
                    .select_key("lines[].product", "product")
                    .select_key("ship_to.country", "country")
                    .select_count()
                    .of_type(ProductInfo)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_6
                results = list(
                    session.query(object_type=Order)
                    .group_by(GroupBy.array("lines[].product"), GroupBy.array("lines[].quantity"))
                    .select_key("lines[].product", "product")
                    .select_key("lines[].quantity", "quantity")
                    .select_count()
                    .of_type(ProductInfo)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_7
                results = list(
                    session.query(object_type=Order)
                    .group_by(GroupBy.array("lines[].product"))
                    .select_key("key()", "products")
                    .select_count()
                    .of_type(ProductsInfo)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_8
                results = list(
                    session.query(object_type=Order)
                    .group_by(GroupBy.array("lines[].product"), GroupBy.field("ship_to.country"))
                    .select_key("lines[].product", "products")
                    .select_key("ship_to.country", "country")
                    .select_count()
                    .of_type(ProductsInfo)
                )
                # endregion

            with store.open_session() as session:
                # region group_by_9
                results = list(
                    session.query(object_type=Order)
                    .group_by(GroupBy.array("lines[].product"), GroupBy.array("lines[].quantity"))
                    .select_key("lines[].product", "products")
                    .select_key("lines[].quantity", "quantities")
                    .select_count()
                    .of_type(ProductsInfo)
                )
                # endregion

File: Documentation/5.4/Samples/php/ClientApi/session/Querying/HowToPerformGroupByQuery.php
<?php

use RavenDB\Documents\Queries\GroupBy;
use RavenDB\Documents\Session\GroupByField;
use RavenDB\Samples\Infrastructure\DocumentStoreHolder;
use RavenDB\Type\StringArray;

class HowToPerformGroupByQuery
{
    public function samples(): void
    {
        $store = DocumentStoreHolder::getStore();
        try {
            $session = $store->openSession();
            try {
                # region group_by_1
                /** @var array<CountryAndQuantity> $orders */
                $orders = $session->query(Order::class)
                    ->groupBy("ShipTo.Country")
                    ->selectKey("ShipTo.Country", "Country")
                    ->selectSum(new GroupByField("Lines[].Quantity", "OrderedQuantity"))
                    ->ofType(CountryAndQuantity::class)
                    ->toList();

                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_2
                $results = $session->query(Order::class)
                    ->groupBy("Employee", "Company")
                    ->selectKey("Employee", "EmployeeIdentifier")
                    ->selectKey("Company")
                    ->selectCount()
                    ->ofType(CountByCompanyAndEmployee::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_3
                /** @var array<CountOfEmployeeAndCompanyPairs> $orders */
                $orders = $session->query(Order::class)
                    ->groupBy("Employee", "Company")
                    ->selectKey("key()", "EmployeeCompanyPair")
                    ->selectCount("Count")
                    ->ofType(CountOfEmployeeAndCompanyPairs::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_4
                /** @var array<ProductsInfo> $products */
                $products = $session->query(Order::class)
                    ->groupBy(GroupBy::array("Lines[].Product"))
                    ->selectKey("key()", "Products")
                    ->selectCount()
                    ->ofType(ProductsInfo::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_5
                /** @var array<ProductInfo> $results */
                $results = $session->advanced()->documentQuery(Order::class)
                    ->groupBy("Lines[].Product", "ShipTo.Country")
                    ->selectKey("Lines[].Product", "Product")
                    ->selectKey("ShipTo.Country", "Country")
                    ->selectCount()
                    ->ofType(ProductInfo::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_6
                /** @var array<ProductInfo> $results */
                $results = $session->query(Order::class)
                    ->groupBy(GroupBy::array("Lines[].Product"), GroupBy::array("Lines[].Quantity"))
                    ->selectKey("Lines[].Product", "Product")
                    ->selectKey("Lines[].Quantity", "Quantity")
                    ->selectCount()
                    ->ofType(ProductInfo::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_7
                /** @var array<ProductsInfo> $results */
                $results = $session->query(Order::class)
                    ->groupBy(GroupBy::array("Lines[].Product"))
                    ->selectKey("key()", "Products")
                    ->selectCount()
                    ->ofType(ProductsInfo::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_8
                /** @var array<ProductsInfo> $results */
                $results = $session->query(Order::class)
                    ->groupBy(GroupBy::array("Lines[].Product"), GroupBy::field("ShipTo.Country"))
                    ->selectKey("Lines[].Product", "Products")
                    ->selectKey("ShipTo.Country", "Country")
                    ->selectCount()
                    ->ofType(ProductsInfo::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region group_by_9
                /** @var array<ProductsInfo> $results */
                $results = $session->query(Order::class)
                    ->groupBy(GroupBy::array("Lines[].Product"), GroupBy::array("Lines[].Quantity"))
                    ->selectKey("Lines[].Product", "Products")
                    ->selectKey("Lines[].Quantity", "Quantities")
                    ->selectCount()
                    ->ofType(ProductsInfo::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

class CountByCompanyAndEmployee
{
    private ?int $count = null;
    private ?string $company = null;
    private ?string $employeeIdentifier = null;

    public function getCount(): ?int
    {
        return $this->count;
    }

    public function setCount(?int $count): void
    {
        $this->count = $count;
    }

    public function getCompany(): ?string
    {
        return $this->company;
    }

    public function setCompany(?string $company): void
    {
        $this->company = $company;
    }

    public function getEmployeeIdentifier(): ?string
    {
        return $this->employeeIdentifier;
    }

    public function setEmployeeIdentifier(?string $employeeIdentifier): void
    {
        $this->employeeIdentifier = $employeeIdentifier;
    }
}

class Order
{

}

class CountryAndQuantity
{
    private ?string $country = null;
    private ?int $orderedQuantity = null;

    public function getCountry(): ?string
    {
        return $this->country;
    }

    public function setCountry(?string $country): void
    {
        $this->country = $country;
    }

    public function getOrderedQuantity(): ?int
    {
        return $this->orderedQuantity;
    }

    public function setOrderedQuantity(?int $orderedQuantity): void
    {
        $this->orderedQuantity = $orderedQuantity;
    }
}

class EmployeeAndCompany
{
    private ?string $employee = null;
    private ?string $company = null;

    public function getEmployee(): ?string
    {
        return $this->employee;
    }

    public function setEmployee(?string $employee): void
    {
        $this->employee = $employee;
    }

    public function getCompany(): ?string
    {
        return $this->company;
    }

    public function setCompany(?string $company): void
    {
        $this->company = $company;
    }
}

class CountOfEmployeeAndCompanyPairs
{
    private ?EmployeeAndCompany $employeeCompanyPair = null;
    private ?int $count = null;

    public function getEmployeeCompanyPair(): ?EmployeeAndCompany
    {
        return $this->employeeCompanyPair;
    }

    public function setEmployeeCompanyPair(?EmployeeAndCompany $employeeCompanyPair): void
    {
        $this->employeeCompanyPair = $employeeCompanyPair;
    }

    public function getCount(): ?int
    {
        return $this->count;
    }

    public function setCount(?int $count): void
    {
        $this->count = $count;
    }
}

class ProductInfo
{
    private ?int $count = null;
    private ?string $product = null;
    private ?int $quantity = null;

    public function getCount(): ?int
    {
        return $this->count;
    }

    public function setCount(?int $count): void
    {
        $this->count = $count;
    }

    public function getProduct(): ?string
    {
        return $this->product;
    }

    public function setProduct(?string $product): void
    {
        $this->product = $product;
    }

    public function getQuantity(): ?int
    {
        return $this->quantity;
    }

    public function setQuantity(?int $quantity): void
    {
        $this->quantity = $quantity;
    }
}

class ProductsInfo
{
    private ?int $count = null;
    private ?StringArray $products = null;
    private ?int $quantity = null;

    public function getCount(): ?int
    {
        return $this->count;
    }

    public function setCount(?int $count): void
    {
        $this->count = $count;
    }

    public function getProducts(): ?StringArray
    {
        return $this->products;
    }

    public function setProducts(?StringArray $products): void
    {
        $this->products = $products;
    }

    public function getQuantity(): ?int
    {
        return $this->quantity;
    }

    public function setQuantity(?int $quantity): void
    {
        $this->quantity = $quantity;
    }

}
